### DISCUSSION TITLE
Ability to add steps dynamically

### DISCUSSION THREAD

**Discussion by eimantaaas** [spatie/laravel-livewire-wizard#14] [Ideas]
How about the ability to add steps dynamically? 👀

For example: at the first step you choose adults count (let's say 2) and then next 2 steps would be for adult information

**Comment by freekmurze** [comment]
I didn't need this myself, so I didn't implement this yet. I've the code for this is very simple, and very well tested and document, I'd considering merging a PR that adds this.

**Comment by nickfls** [comment]
Can you at least point to a direction where to look?
I imagine this would involve an event emitted upon step change, but cannot figure it out..

**Reply by ju5t** [reply]
Emitting an event could work. I think all you need to know is everything revolves around `stepNames()`. Just make sure it's available there, in the right order. This isn't supported yet, so it would require changing the current code.

**Reply by nickfls** [reply]
I am able to _create_ dynamic steps, but no _replace_ them.
Wizard component has a flag (model id) that is empty on load, so i adjust listeners and `steps()`:

```php
    protected function getListeners()
    {
        return collect(parent::getListeners())
            ->put('model:created', 'setModelID')
            ->toArray();
    }

    public function steps(): array
    {
        return collect([])
            ->unless($this->modelID, fn($steps) => $steps->push(CreateModelComponent::class))
            ->when($this->modelID, fn($steps) => $steps->push(UpdateModelComponent::class))
            ->when($this->modelID, fn($steps) => $steps->push(ReviewComponent::class))
            ->toArray();
    }
```
and `emitUp('model:created')` from `CreateModelComponent`

But this would not work, obviously, as that would change livewire checksum and the ID cannot be found.

So, i think, _in this situation_, to approach it is either have two components and once i have the ID, redirect to a wizard 
_OR_
UpsertModelComponent that replaces the view based on presence of modelID:

```php
    public function render()
    {
        $view = $this->state()->get('wizard:upsert-model.modelID') ? 'update' : 'create';

        return view("livewire.model-wizard.steps.model-$view", [
            'model' => $this->state()->model(),
        ]);
    }
```

more or less, i imagine.
The more i think about it, the less i like upsert and the more i like redirect to a wizard that gets a model at all times

**Reply by TiesPol** [reply]
Hey @nickfls, did you came up with a solution for this? I'm trying to build this myself as well.
